pinephone/vibrator/FFDevice.cpp

```diff
@@ -0,0 +1,135 @@
+            return true;
+        }
+        return false;
+    };
+    if (recursive_dir_observer(input_path, observe_func, &event_path)) {
+        event_path = "/dev/input/" + event_path;
+        auto device = std::unique_ptr<FFDevice>(new FFDevice());
+        device->event_fd = open(event_path.c_str(), O_RDWR);
+        if (device->event_fd == -1) {
+            LOG(ERROR) << "FFDevice: error opening event file";
+            return nullptr;
+        }
+        return device;
+    }
+    LOG(ERROR) << "FFDevice: event path not found";
+    return nullptr;
+}
+
+FFDevice::~FFDevice() {
+    if (event_fd >= 0) {
+        close(event_fd);
+    }
+}
+
+void FFDevice::vibrate(int duration_ms) {
+    if (last_effect_id != -1) {
+        off();
+    }
+
+    /* Prepare effect */
+    ff_effect effect = {.type = FF_RUMBLE,
+                        .u.rumble.strong_magnitude = 65535,
+                        .u.rumble.weak_magnitude = 65535,
+                        .replay.length = static_cast<__u16>(duration_ms),
+                        .replay.delay = 0,
+                        .id = -1 /* set to -1 for allocate new effect */};
+
+    /* Send the effect to the driver */
+    if (ioctl(event_fd, EVIOCSFF, &effect) < 0) {
+        LOG(ERROR) << "FFDevice: send effect failed";
+        return;
+    }
+    last_effect_id = effect.id;
+
+    /* Prepare the play event */
+    input_event play = {
+        .type = EV_FF, .code = static_cast<__u16>(effect.id), .value = 1};
+
+    /* Playing the effect in the driver */
+    if (write(event_fd, &play, sizeof(play)) == -1) {
+        LOG(ERROR) << "FFDevice: play effect failed";
+    }
+}
+
+void FFDevice::off() {
+    if (last_effect_id == -1) {
+        return;
+    }
+
+    /* Cleanup */
+    if (ioctl(event_fd, EVIOCRMFF, last_effect_id) < 0) {
+        LOG(ERROR) << "FFDevice: remove effect failed";
+    }
+    last_effect_id = -1;
+}
+
+}  // namespace vibrator
+}  // namespace hardware
+}  // namespace android
+}  // namespace aidl
+
+/* Recursive traversal of the directory tree */
+static bool recursive_dir_observer(const char *path, dirfunc_t func,
+                                   void *param) {
+    DIR *d = opendir(path);
+    if (d == NULL) {
+        return false;
+    }
+
+    struct dirent *dir;  /* for the directory entries */
+    bool finded = false; /* to exit recursion */
+    std::list<std::string> paths;
+    while ((dir = readdir(d)) !=
+           NULL) /* if we were able to read somehting from the directory */
+    {
+        if (dir->d_type == DT_DIR && strcmp(dir->d_name, ".") != 0 &&
+            strcmp(dir->d_name, "..") != 0) /* if it is a directory */
+        {
+            /* call observe function */
+            if (func(dir->d_name, param)) {
+                finded = true;
+                break;
+            }
+            /* save directory path for future observation */
+            std::string dirpath = path + std::string("/") + dir->d_name;
```

Comment:
Consider replacing this with a more efficient version with `std::ostringstream`:
```
std::ostringstream dirpath << path << "/" << dir->d_name;
paths.push_back(dirpath.str());
```
This would avoid unnecessary temporary `std::string` allocations.

Reply:
Done.